### PR TITLE
[Feature] Admin Prereg Upgrade

### PR DESCRIPTION
**Pull request by Ghalko**
## Purpose

Replaces #5533 

## Changes

Adding features for better prereg use.

## Side effects

None.

## Ticket

OSF-6260, OSF-6261, OSF-6262, OSF-6263, OSF-6264

### CODE REVIEW

File: website/project/model.py [hunk 1]
```diff
@@ -3815,6 +3815,8 @@ class DraftRegistration(StoredObject):
 
     approval = fields.ForeignField('draftregistrationapproval', default=None)
 
+    status_logs = fields.DictionaryField(default=dict)
```
Comment (sloria) [inline website/project/model.py]: Please document the expected format for this field.

File: website/project/model.py [hunk 2]
```diff
@@ -3879,6 +3881,13 @@ def is_rejected(self):
         else:
             return False
 
+    @property
+    def logs(self):
+        return [
+            '{} on {}'.format(v.title(), k)
```
Comment (sloria) [inline website/project/model.py]: Try to separate presentation/formatting code from the model. This should probably be a helper function, e.g. `format_status_logs(draft)`.

File: website/project/model.py [hunk 3]
```diff
@@ -3918,13 +3927,19 @@ def update_metadata(self, metadata):
         self.registration_metadata.update(metadata)
         return changes
 
+    def add_status_log(self, action):
+        self.status_logs.update({
+            datetime.datetime.utcnow().strftime('%Y-%m-%d %H:%M'): action
```
Comment (sloria) [inline website/project/model.py]: Why are status logs represented as a dictionary keyed by date? Why optimize for accessing by a single date? It might make sense to follow the same modeling as NodeLog/Node.logs.

File: website/static/js/registrationUtils.js [hunk 4]
```diff
@@ -11,7 +11,7 @@ var bootbox = require('bootbox');
 var moment = require('moment');
 var History = require('exports?History!history');
 
-require('js/koHelpers');
+var koHelpers = require('js/koHelpers');
```
Comment (sloria) [inline website/static/js/registrationUtils.js]: This var isn't used anywhere. We only require `koHelpers` in order to register our custom binding handlers.

**Comment by sloria** [comment]
Pass finished.

File: website/project/model.py (hunk 3 above)
Comment (Ghalko) [inline website/project/model.py]: This is a really simple log for David just to be able to see how many stages one of these went through. It would certainly be possible to make a new model of a stripped down log, but I thought that it would be simpler to have a list, however a list really isn't supported in modm beyond using other models.

File: website/static/js/registrationUtils.js (hunk 4 above)
Comment (Ghalko) [inline website/static/js/registrationUtils.js]: Removed.

File: website/project/model.py (hunk 2 above)
Comment (Ghalko) [inline website/project/model.py]: This hinges on the answer to https://github.com/CenterForOpenScience/osf.io/pull/5644#discussion_r62885659

File: website/project/model.py (hunk 1 above)
Comment (Ghalko) [inline website/project/model.py]: This hinges on the answer to https://github.com/CenterForOpenScience/osf.io/pull/5644#discussion_r62885659

**Comment by Ghalko** [comment]
@sloria should I make a new model just to keep these simple logs?

**Comment by sloria** [comment]
A new model would make sense.

File: website/project/model.py (hunk 2 above)
Comment (Ghalko) [inline website/project/model.py]: Replaced with serializer in prereg admin.

File: admin/pre_reg/serializers.py [hunk 5]
```diff
@@ -21,18 +21,16 @@ def serialize_user(user):
...
-    registration_choice = draft.approval.meta.get('registration_choice', None)
-    if registration_choice == EMBARGO:
-        time = parser.parse(draft.approval.meta['embargo_end_date'])
-        embargo = iso8601format(time) if json_safe else time
+    node_url = get_url(draft)
+
+    embargo = get_embargo(draft, json_safe)
+
+    if draft.is_approved:
+        status = get_approval_status(draft)
```
Comment (mfraezz) [inline admin/pre_reg/serializers.py]: Minor, but why not have this helper determine approval status in the other cases as well?

Comment (Ghalko) [inline admin/pre_reg/serializers.py]: Done.

File: admin/pre_reg/utils.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    'n_initiator': '-initiator',
+    'title': 'title',
+    'n_title': '-title',
+    'date': 'date',
+    'n_date': '-date',
+    'state': 'state',
+    'n_state': '-state',
+}
+
+VIEW_STATUS = {
```
Comment (mfraezz) [inline admin/pre_reg/utils.py]: This appears unused. It's included in the context for `DraftListView`, but is unused by the template.

File: website/project/model.py [hunk 7]
```diff
@@ -3782,6 +3782,30 @@ def to_json(self):
...
+    @classmethod
+    def add_log(cls, action):
+        log = cls(
+            action=action,
+        )
+        log.save()
+        return log
+
+    def __repr__(self):
+        return 'DraftRegistrationLog: {}'.format(self._id)
```
Comment (sloria) [inline website/project/model.py]: This is ok, but it should be consistent with other `__repr__` methods. It would also be helpful to include the date and action.

It should look something like

```
<DraftRegistrationLog(date=datetime.datetime(...), action='foo') with _id 'abc12'>
```

File: website/project/model.py [hunk 8]
```diff
@@ -3782,6 +3782,30 @@ def to_json(self):
...
+    field - _id - primary key
+    field - date - date of the action took place
+    field - action - simple action to track what happened
+    """
+    _id = fields.StringField(primary=True, default=lambda: str(ObjectId()))
+    date = fields.DateTimeField(default=datetime.datetime.utcnow, index=True)
+    action = fields.StringField(index=True)
+
+    @classmethod
+    def add_log(cls, action):
```
Comment (sloria) [inline website/project/model.py]: This doesn't seem necessary, and it's confusingly named; you don't add logs to a DraftRegistrationLog.  It would make more sense to have `DraftRegistration#add_status_log`.

File: website/project/model.py [hunk 9]
```diff
@@ -3782,6 +3782,30 @@ def to_json(self):
...
+class DraftRegistrationLog(StoredObject):
+    """ Simple log to show status changes for DraftRegistrations
+
+    field - _id - primary key
+    field - date - date of the action took place
+    field - action - simple action to track what happened
+    """
+    _id = fields.StringField(primary=True, default=lambda: str(ObjectId()))
+    date = fields.DateTimeField(default=datetime.datetime.utcnow, index=True)
+    action = fields.StringField(index=True)
```
Comment (sloria) [inline website/project/model.py]: It would be helpful to have constants for the valid values of `action` (similar to `NodeLog`).

File: admin/pre_reg/views.py [hunk 10]
```diff
@@ -17,19 +18,18 @@
...
 
 class DraftListView(PreregAdmin, ListView):
     template_name = 'pre_reg/draft_list.html'
-    paginate_by = 10
-    paginate_orphans = 1
-    ordering = '-approval.initiation_date'
+    ordering = '-date'
     context_object_name = 'draft'
 
     def get_queryset(self):
```
Comment (sloria) [inline admin/pre_reg/views.py]: I know the below code is from a previous PR, but there is a helper function for getting the prereg metaschema: `website.prereg.utils.get_prereg_schema`.

File: admin/pre_reg/utils.py [hunk 11]
```diff
@@ -0,0 +1,36 @@
+
+SORT_BY = {
```
Comment (sloria) [inline admin/pre_reg/utils.py]: Why is this necessary? Couldn't the client just pass `-initiator` instead of `n_initiator`?

File: admin/pre_reg/views.py [hunk 12]
```diff
@@ -54,8 +63,42 @@ def get_context_data(self, **kwargs):
                 for d in query_set
             ],
             'page': page,
+            'p': self.get_paginate_by(query_set),
+            'SORT_BY': SORT_BY,
+            'order': self.get_ordering(),
+            'status': self.request.GET.get('status', 'all'),
         }
 
+    def get_paginate_by(self, queryset):
```
Comment (sloria) [inline admin/pre_reg/views.py]: The `queryset` param is unused.

File: website/project/model.py [hunk 13]
```diff
@@ -3782,6 +3782,30 @@ def to_json(self):
             "text": self.text
         }
 
+
+class DraftRegistrationLog(StoredObject):
+    """ Simple log to show status changes for DraftRegistrations
+
+    field - _id - primary key
+    field - date - date of the action took place
+    field - action - simple action to track what happened
```
Comment (sloria) [inline website/project/model.py]: It might also make sense to store the user who generated the logged action.

**Comment by sloria** [comment]
Pass finished.

File: website/project/model.py (hunk 7 above)
Comment (Ghalko) [inline website/project/model.py]: Done.

File: website/project/model.py (hunk 8 above)
Comment (Ghalko) [inline website/project/model.py]: Done.

File: website/project/model.py (hunk 9 above)
Comment (Ghalko) [inline website/project/model.py]: Done.

File: admin/pre_reg/views.py (hunk 10 above)
Comment (Ghalko) [inline admin/pre_reg/views.py]: Done.

File: admin/pre_reg/views.py (hunk 12 above)
Comment (Ghalko) [inline admin/pre_reg/views.py]: It is a default in Django.

File: admin/pre_reg/utils.py (hunk 11 above)
Comment (Ghalko) [inline admin/pre_reg/utils.py]: This is a consolidation so that the template renderer doesn't need to deal with `-` since this dict is used in sorting the lists.

File: website/project/model.py (hunk 13 above)
Comment (Ghalko) [inline website/project/model.py]: Added

**Comment by Ghalko** [comment]
@sloria This is ready for final pass/merging.

File: website/project/model.py [hunk 14]
```diff
@@ -3815,6 +3841,8 @@ class DraftRegistration(StoredObject):
 
     approval = fields.ForeignField('draftregistrationapproval', default=None)
 
+    status_logs = fields.ForeignField('draftregistrationlog', list=True)
```
Comment (sloria) [inline website/project/model.py]: Rather than DraftReg having a forward-ref to DraftRegLog, why not flip the relationship (Log-> DraftReg)? This would be consistent with NodeLog -> Node.

**Comment by sloria** [comment]
Pass finished. Just one suggestion: https://github.com/CenterForOpenScience/osf.io/pull/5644#discussion_r64462912

File: website/project/model.py (hunk 14 above)
Comment (Ghalko) [inline website/project/model.py]: Consistent, yes, however, what is the general rationale for this architecture?
- Does having a list of logs imply a new set of automatic queries? If so, why do we even have this option? In what case would this ever be optimal?
- Why is a backref, assuming the previous question is reasonable, any better? Won't it load the DraftRegistration with each log loaded?
-

Comment (mfraezz) [inline website/project/model.py]: >Does having a list of logs imply a new set of automatic queries? 

Assuming I understand your question, you could still have a property on this model that just returns `DraftRegistrationLog.find(Q('owner', 'eq', self._id))` or similar. You probably shouldn't need to change `'logs': map(serialize_draft_logs, draft.status_logs)` above.

> If so, why do we even have this option? In what case would this ever be optimal?

In the case where there exist a large number of DRLogs, you gain the benefit of not needing to load each into memory when loading the DraftRegistration.

> Why is a backref, assuming the previous question is reasonable, any better? 

It's not a *backref*, exactly. It's a forward ref in the other direction.

> Won't it load the DraftRegistration with each log loaded?

Yes, which is better than loading every DRLog when the associated DraftRegistration is loaded -- one reason is above. Another reason is that the DR will likely be cached already if you're loading a DRLog; the reverse does not necessarily hold true.

**Comment by sloria** [comment]
Merged into `develop`. 👍
